lib/tracing-limit/src/lib.rs

```diff
@@ -10,165 +10,200 @@ use std::{
+#[derive(Eq, PartialEq, Hash)]
+struct RateKeyIdentifier {
+    callsite: Identifier,
+    rate_limit_key: Option<String>,
+    component_name: Option<String>,
 }
 
-#[derive(Debug)]
-struct State {
-    start: Option<Instant>,
-    count: AtomicUsize,
-    limit: u64,
-    message: String,
-}
-
-impl Limit {
-    fn create_event<S: Subscriber>(
-        &self,
-        id: &Identifier,
-        ctx: &Context<S>,
-        message: String,
-        rate_limit: u64,
-    ) {
-        let store = self.callsite_store.read().unwrap();
-        let metadata = store.get(id).unwrap();
-
-        let fields = metadata.fields();
-
-        let message = display(message);
-
-        if let Some(message_field) = fields.field("message") {
-            let values = [(&message_field, Some(&message as &dyn Value))];
+pub struct RateLimitedLayer<S, L>
+where
+    L: Layer<S> + Sized,
+    S: Subscriber,
+{
+    events: RwLock<HashMap<RateKeyIdentifier, State>>,
+    inner: L,
 
-            let valueset = fields.value_set(&values);
-            let event = Event::new(metadata, &valueset);
-            ctx.event(&event);
-            drop(store);
-        } else {
-            let values = [(
-                &fields.field(RATE_LIMIT_FIELD).unwrap(),
-                Some(&rate_limit as &dyn Value),
-            )];
+    // TODO is this right?
+    _subscriber: std::marker::PhantomData<S>,
+}
 
-            let valueset = fields.value_set(&values);
-            let event = Event::new(metadata, &valueset);
-            ctx.event(&event);
-            drop(store);
+impl<S, L> RateLimitedLayer<S, L>
+where
+    L: Layer<S> + Sized,
+    S: Subscriber,
+{
+    pub fn new(layer: L) -> Self {
+        RateLimitedLayer {
+            events: RwLock::<HashMap<RateKeyIdentifier, State>>::default(),
+            inner: layer,
+            _subscriber: std::marker::PhantomData,
         }
     }
 }
 
-impl<S> Layer<S> for Limit
+impl<S, L> Layer<S> for RateLimitedLayer<S, L>
 where
-    S: Subscriber,
-    Self: 'static,
+    L: Layer<S>,
+    S: Subscriber + for<'a> tracing_subscriber::registry::LookupSpan<'a>,
 {
     fn register_callsite(&self, metadata: &'static Metadata<'static>) -> Interest {
+        let inner = self.inner.register_callsite(metadata);
+        if inner.is_never() {
+            // if wrapped layer doesn't care about the event, don't bother rate limiting
+            return inner;
+        }
+
         if metadata
             .fields()
             .iter()
-            .any(|f| f.name() == RATE_LIMIT_FIELD)
+            .any(|f| f.name() == RATE_LIMIT_SECS_FIELD)
         {
-            let id = metadata.callsite();
-            let mut events = self.events.write().expect("lock poisoned!");
-
-            let state = State {
-                start: None,
-                count: AtomicUsize::new(0),
-                limit: DEFAULT_LIMIT,
-                message: String::new(),
-            };
+            Interest::sometimes()
```

Comment:
This is a minor issue and I'm not sure if I'm correct, `Interest::sometimes` means: `enabled` will be called on each event, if it returns true then `on_event` will be called, otherwise don't call it. 

Because this layer just calls `inner.enabled` and doesn't do anything else, it should also return `inner.register_callsite` here.

Reply:
Good catch, you are right, this should just proxy to inner.